core/invoker/src/main/scala/whisk/core/invoker/Invoker.scala

```diff
@@ -85,6 +86,41 @@ object Invoker {
       abort()
     }
 
+    val proposedInvokerId: Option[Int] = args.headOption.map(_.toInt)
+    val assignedInvokerId = proposedInvokerId
+      .map { id =>
+        logger.info(this, s"invokerReg: using proposedInvokerId ${id}")
+        id
+      }
+      .getOrElse {
+        val invokerName = config.invokerName
```

Comment:
Can invokerName come from an separate environment variable or arg? I think it would be possible from typesafe config, but don't think so with whisk Config. In our case, since the ID is an arg we can generate it with a script, but it cannot be the IP address directly since it is not an Int, or easily translated to an Int that is "small".  It would be better to use the invokerName, and leverage your approach, but we cannot do that with `config.invokerName` afaik without designating a name unique to each instance. 

This may be more of a problem of using IndexedSeq[Int] in the controller with padding where a large value for proposedInvokerId will cause havoc due to padding of invoker registration for each id lower than the proposed. 

It would be great to document this treatment of instanceid (I couldn't find any) to let devs know that using "any Int" is not acceptable; or else it would be good to have a workaround, I may be doing it wrong - we hoped to use an int derived from IP number, but the result is that, for example, the controller/invokers endpoint displays as massive quantity of invokers, all of which are offline, due to the padding. I'm not sure if it would be possible to change this type from Int to String in controllers invoker registration logic, but that would be one way to solve it, if using a referenced env var (e.g. `INVOKER_NAME=${SOME_VAR}`)  for invoker name is not possible.

Reply:
If I am reading the pr correctly that name and ordinal assigned as the id for load balancing are completely decoupled. The name could be any value/string. The increments through redis make sure you have a monotoicnally increasing value.

Reply:
My question is about whether `config.invokerName` has to be statically assigned as an environment variable `INVOKER_NAME=MyInvokerName`, or if it can be derived from either a system property or other environment variable, e.g. `INVOKER_NAME=${SOME_VALUE}` (where `SOME_VALUE` is another variable). Or, if it was possible to pass it as an arg, that would also be a way to use `$SOME_VALUE`, but currently it isn't.

Reply:
Ah. Thanks for the clarification. My understanding is as long as you can generate a stable name (across restarts) it should no longer matter. The reason the name has to be stable is for the redis negotiation. Every new name would bump the count otherwise. So I think the naming could be generalized accordingly.

Reply:
I was wondering fi a separate redis client to get the Invoker Id is an overkill.  Isn't just keeping a circular buffer of " InvokerIds& Names" in the Controller sufficient to assign a new id to the Invoker.  the same Circular buffer can log data from the Health stats and registration status. "Number of Invokers" per Controller can be set in Limits and higher-level scaling with HA replicated controllers.

Reply:
The redis instance is already deployed (for api gw) and I expect will have more uses as we evolve the control/data plane.

The proposed patch here is preserving current invariants which doesn’t require the loadbalancer to change the way the topics are mapped to invokers. But it breaks the naming and topic conflation which exists otherwise. I expect a future iteration to go further and delegate topic assignment to the loadbalancer on demand/as invokers register. This could even permit work stealing, with some imagination.

Reply:
Rabbah, Thank you for the clarification. So the idea of Redis is so that the Openwhisk configuration(/and)state can be reloaded, incase the controller goes down and comes back up?

Reply:
Here, the concern is primarily with the invokers going down and then recovering. Assuming a stable id, the invoker will pick up its previous active topic. As a result, there are no orphaned topics when an invoker recovers. I think @dgrove-oss has described somewhere a possible algorithm for back filing topics through the loadbalancer (where then it's closer to what you're alluding to in your previous comment).

Reply:
If  "stable id" == "same id as before going down", then I guess the Invoker needs a way to Identify itself ( similar  to a MAC/Ethernet Id that gets assigned an IP address on ARP resolution ),  when it comes back up; But since functions ( Invoker actions) are stateless by nature, we can just rerout(In kafka terminology I guess re-topic) the orhaned actions, instead of putting the burden on the Invoker to get its previous name - Could u point me to the algorithm that u r referring to.

Reply:
Baby steps - this is an non-invasive pr in that it preserves the current behavior but breaks the bijection between invoker name and topic id. @dgrove-oss will know better where he posted it, maybe the dev list. My recollection: the lb can detect an orphaned topic and assign it to a new invoker once it registers itself.

Reply:
OK,  makes sense. Thank you!, I will syncup w/ dgrove-oss et al  on the follow-up items from this PR, if any.

Reply:
@tysonnorris re config.invokerName, this is being read from the environment variable INVOKER_NAME  (@rabbah had asked me to not read sys.env directly, but instead access the container's environment through the WhiskConfig object).   Does this work for your scenario?

Reply:
@ServoKvd, I think I failed to actually post something about possible followons to this PR.  As Rodric said, this PR was meant as a small step to allow us to decouple the invoker name from the kafka topic that the invoker is assigned to service.  Useful for some of the kube work and to give some flexibility in deployment.  In the future the decoupling could allow a fancier load balancer that could more dynamically assign invokers to topics and also support an elastic pool of invokers that could grow and shrink depending on overall system load.  All details still to be worked out 😁

Reply:
@dgrove-oss unfortunately setting INVOKER_NAME doesn't work in our case, since the env value will be the same for each instance (when running via mesos/marathon). If it was possible to reference *another* env value as the value for INVOKER_NAME, it would work, e.g. `INVOKER_NAME=${SOME_OTHER_ENV}`, since there will be a few of these values set per instance automatically (but cannot be used to explicitly set some var unique per instance). 

I think a bigger change would be required to support this for our case, like either use TypeSafe Config (instead of whisk Config), or arg flags (e.g. you can specify either id OR name as an arg), or some extension for looking up the name that may be appropriate for particular environments (e.g. in mesos we could lookup a name from some other env vars; other envs may have other ways of associating names, etc).

Reply:
@tysonnorris thanks; I understand the mesos/marathon constraints now.  They are different than kube 😞  Would it be acceptable for me to generalize in a followup PR? I have to admit I'd like to get this one merged without more iterations so we can unblock the DaemonSet PR on the kube deploy project.

I don't think it would be that hard for me to do a followup PR to allow a combination of arg flags and environment variables to be used.

Reply:
@dgrove-oss thanks, I understand. We can work with the current requirement to assign instance id, it is just very clunky :) If you get to a followup for the other combinations I'd be happy to test it.